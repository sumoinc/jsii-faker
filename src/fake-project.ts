import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";

export class FakeProject extends TypeScriptProject {
  constructor(options: TypeScriptProjectOptions) {
    super(options);
  }
}
