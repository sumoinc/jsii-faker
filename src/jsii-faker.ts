import * as spec from "@jsii/spec";
import { Component, JsonFile } from "projen";
import { TypeScriptProject } from "projen/lib/typescript";
import { ValueOf } from "type-fest";

/**
 * !!! WARNING - VERY EXPERIMENTAL !!!
 *
 * This is possibly a terrible idea, but for now it's the only way that I can
 * find to generate a minimal `.jsii` file so that Projen can then consume this
 * package when called similar to:
 *
 *  `npx projen new --from @my-scope/my-package my-project-type`
 *
 * What you are missing by not using the projen built in JsiiProject type:
 *
 * - No automatic API.md file.
 * - No compatable versions in other language like GO, Java, Python.
 * - Other things?
 *
 * What you gain by using this:
 *
 * - Freedom from some fairly significant limitations that JSII imposes on your
 *   Typescript (https://aws.github.io/jsii/user-guides/lib-author/typescript-restrictions/)
 * - Ability to publish / consume pure Typescript project types more easily.
 *
 * JSII CODE
 *
 * See the following for more info on how jsii proper assembles and validates the file:
 * https://github.com/aws/jsii/blob/62d2d37212a111f4adc63998245d280f1c43ff86/packages/jsii/lib/assembler.ts
 *
 */

const ProjenBaseFqn = {
  TYPESCRIPT_PROJECT: "projen.typescript.TypeScriptProject",
  TYPESCRIPT_PROJECT_OPTIONS: "projen.typescript.TypeScriptProjectOptions",
} as const;

export interface ClassTypeOptions {
  /**
   * The name of the class.
   *
   * @example "MyProject"
   */
  name: string;
  /**
   * The FQN for the base class this class is extending.
   *
   * @default ProjenBaseFqn.TYPESCRIPT_PROJECT
   */
  baseFqn?: ValueOf<typeof ProjenBaseFqn> | string;
  /**
   * The FQN for the options for this class.
   *
   * @default ProjenBaseFqn.TYPESCRIPT_PROJECT_OPTIONS
   */
  optionsFqn?: ValueOf<typeof ProjenBaseFqn> | string;
}

export class JsiiFaker extends Component {
  // find project singleton
  public static of(project: TypeScriptProject): JsiiFaker | undefined {
    const isDefined = (c: Component): c is JsiiFaker => c instanceof JsiiFaker;
    return project.components.find(isDefined);
  }

  private _assemblyName: string;
  private _types: { [name: string]: spec.Type } = {};

  constructor(public readonly project: TypeScriptProject) {
    super(project);

    /**
     * In JSII, the assembly name is essentially the package name. It's used as a
     * scope when targeting types and metadata in other "jsii assemblies" that
     * might be in sub-packages used by the project.
     *
     * For this case, we'll just use the pachage name from Projen.
     */
    this._assemblyName = this.project.package.packageName;

    new JsonFile(project, ".jsii", {
      obj: () => {
        return {
          name: this._assemblyName,
          types: this._types,
        };
      },
    });
  }

  public toJSON = () => {
    return {
      types: this._types,
    };
  };

  public addClassType(options: ClassTypeOptions) {
    const fqn = [this._assemblyName, options.name].join(".");
    const type: spec.ClassType = {
      assembly: this._assemblyName,
      base: options.baseFqn ?? ProjenBaseFqn.TYPESCRIPT_PROJECT,
      fqn,
      kind: spec.TypeKind.Class,
      name: options.name,
      initializer: {
        parameters: [
          {
            name: "options",
            type: {
              fqn:
                options.optionsFqn ?? ProjenBaseFqn.TYPESCRIPT_PROJECT_OPTIONS,
            },
          },
        ],
      },
    };

    this._types[fqn] = type;
  }
}
