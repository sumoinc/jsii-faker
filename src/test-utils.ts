import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";

/**
 * Helper to generate and return files.
 * Useful when trying to test the final output of generated files.
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce(
      (obj, key) => {
        return {
          ...obj,
          [key]: snapshot[key],
        };
      },
      {} as { [key: string]: any },
    );
  return filtered;
};
