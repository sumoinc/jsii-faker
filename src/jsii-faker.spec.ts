import * as spec from "@jsii/spec";
import { FakeProject } from "./fake-project";
import { JsiiFaker } from "./jsii-faker";
import { synthFiles } from "./test-utils";

describe("Success conditions", () => {
  test("Smoke Test", () => {
    const { faker } = buildFakeProject();
    expect(faker).toBeTruthy();
  });
});

describe("Files", () => {
  test("Smoke test", () => {
    const { project } = buildFakeProject();

    const content = synthFiles(project);
    expect(content[".jsii"]).toBeTruthy();
    expect(content[".jsii"]).toMatchSnapshot();

    //console.log(content[".jsii"]);
  });

  test("Assembly name should match package name", () => {
    const { project } = buildFakeProject();

    const files = synthFiles(project);
    const content = JSON.parse(files[".jsii"]) as spec.Assembly;
    expect(content.name).toBe("@my-scope/myproject");

    //console.log(content);
  });
});

/**
 * Testing helper
 */
const buildFakeProject = () => {
  const project = new FakeProject({
    name: "my-project",
    packageName: "@my-scope/myproject",
    defaultReleaseBranch: "main",
  });

  const faker = new JsiiFaker(project);
  faker.addClassType({
    name: "FooProject",
  });
  return { project, faker };
};
