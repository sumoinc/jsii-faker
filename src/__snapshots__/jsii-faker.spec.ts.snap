// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files Smoke test 1`] = `
"{
  "name": "@my-scope/myproject",
  "types": {
    "@my-scope/myproject.FooProject": {
      "assembly": "@my-scope/myproject",
      "base": "projen.typescript.TypeScriptProject",
      "fqn": "@my-scope/myproject.FooProject",
      "kind": "class",
      "name": "FooProject",
      "initializer": {
        "parameters": [
          {
            "name": "options",
            "type": {
              "fqn": "projen.typescript.TypeScriptProjectOptions"
            }
          }
        ]
      }
    }
  },
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\"."
}
"
`;
